[inductor][fx pass] handle numpy compatibility arg names

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113078

Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

davidberard98 committed:
[inductor][fx pass] handle numpy compatibility arg names

the "dim" kwarg can also be referred to with "axis". we need to handle this case, otherwise we assume that "dim=0" even if "axis=1" was set.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51045878).

davidberard98 committed:
Update on "[inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -1353,11 +1353,28 @@ def run_node(self, old_node):
-        else node.kwargs.get(kwarg_name)
-    )
+    # see python_arg_parser.cpp; e.g. kwargs named "dim" also match "axis"
+    numpy_compat = {
```
Can this just be a pass on pre-grad stack ? these args get canonicalized in torch dispatch

davidberard98 replied:
@eellison is there a way for me to set the priority of the pass so that we can guarantee that the numpy-arg-canonicalization pass will run before the split-cat pass?

yanboliang replied:
You can follow the current ```normalization_pass```:
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/split_cat.py#L165

As @eellison mentioned, only pre_grad fx passes suffer from this issue, we just need to handle them.

davidberard98 committed:
Update on "[inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

eellison commented:
take out of draft and mark `ready for review` when it's ready!

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51045878).

davidberard98 committed:
Update on "[WIP][inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP][inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51045878](https://our.internmc.facebook.com/intern/diff/D51045878)

[ghstack-poisoned]

davidberard98 commented:
seems like no compile time regression

https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2007%20Nov%202023%2023%3A30%3A02%20GMT&stopTime=Tue%2C%2014%20Nov%202023%2023%3A30%3A02%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/davidberard98/237/orig&lCommit=eef75bf5021870d04cd684ef599e4522b8dcf07a&rBranch=main&rCommit=a1e3c501652101e8b37baac62216db7ca22c9923

<img width="724" alt="Screenshot 2023-11-14 at 3 31 10 PM" src="https://github.com/pytorch/pytorch/assets/5067123/a8934b38-d216-492b-b14c-3cdc0021d115">

davidberard98 commented:
cc @mengluy0125 @jackiexu1992 if you add pre-grad fx passes on methods you would probably need to add handling for those as well (e.g. `x.sum(axis=1)`)

eellison commented on torch/_inductor/fx_passes/misc_patterns.py:
```diff
@@ -62,3 +66,70 @@ def randperm_index_replacement(x, slice_shape):
+                            replaceable_kwargs.update(self.numpy_compat[param_name])
+
+                self.cache[node.target] = replaceable_kwargs
+
```
continue here if `not replaceable_kwargs` ?

eellison reviewed (approved):
nice !

davidberard98 committed:
Update on "[inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor][fx pass] handle numpy compatibility arg names"


Fixes #113038 

the "dim" kwarg can also be referred to with "axis" - handle this case.

https://github.com/pytorch/pytorch/blob/21b6030ac3e6461c914e18f9ed60f6f713b3133b/torch/csrc/utils/python_arg_parser.cpp#L72-L77

previously, if the "axis" kwarg was used, it would not be matched and "dim" would default to 0.

https://github.com/pytorch/pytorch/blob/75adb9f37150a670e88c82a2090fe9e40a5602c6/torch/_inductor/fx_passes/split_cat.py#L172-L176

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
it's merged, so I will manually close the issue
